File: agents/app/service_auth.py
from __future__ import annotations

import os

from fastapi import HTTPException, Request, status
from google.auth.transport.requests import Request as GoogleAuthRequest
from google.oauth2 import id_token

_google_request = GoogleAuthRequest()


def _is_truthy(value: str | None) -> bool:
    return str(value or "").strip().lower() in {"1", "true", "yes", "on"}


def _is_falsey(value: str | None) -> bool:
    return str(value or "").strip().lower() in {"0", "false", "no", "off"}


def should_require_service_auth() -> bool:
    flag = os.getenv("SERVICE_TO_SERVICE_AUTH")
    if flag and _is_falsey(flag):
        return False
    if flag and _is_truthy(flag):
        return True
    return bool(os.getenv("K_SERVICE"))


def expected_audience(request: Request) -> str | None:
    forwarded_proto = request.headers.get("x-forwarded-proto")
    protocol = (forwarded_proto.split(",")[0].strip() if forwarded_proto else request.url.scheme) or "http"
    host = request.headers.get("host")
    if not host:
        return None
    return f"{protocol}://{host.split(',')[0].strip()}"


def require_service_caller(request: Request) -> None:
    if not should_require_service_auth() or request.url.path == "/healthz":
        return

    authorization = request.headers.get("authorization", "")
    scheme, _, token = authorization.partition(" ")
    if scheme.lower() != "bearer" or not token.strip():
        raise HTTPException(
            status_code=status.HTTP_401_UNAUTHORIZED,
            detail="Missing service identity",
        )

    audience = expected_audience(request)
    if not audience:
        raise HTTPException(
            status_code=status.HTTP_401_UNAUTHORIZED,
            detail="Missing service audience",
        )

    try:
        id_token.verify_oauth2_token(token.strip(), _google_request, audience=audience)
    except Exception as exc:  # pragma: no cover - service boundary
        raise HTTPException(
            status_code=status.HTTP_401_UNAUTHORIZED,
            detail="Invalid service identity",
        ) from exc

File: backend/src/auth/serviceAuth.js
import { OAuth2Client } from "google-auth-library";

const authClient = new OAuth2Client();

function createHttpError(status, message) {
  const error = new Error(message);
  error.statusCode = status;
  return error;
}

function isTruthy(value) {
  return ["1", "true", "yes", "on"].includes(String(value).trim().toLowerCase());
}

function isFalsey(value) {
  return ["0", "false", "no", "off"].includes(String(value).trim().toLowerCase());
}

function shouldRequireServiceAuth() {
  if (process.env.SERVICE_TO_SERVICE_AUTH && isFalsey(process.env.SERVICE_TO_SERVICE_AUTH)) {
    return false;
  }

  if (process.env.SERVICE_TO_SERVICE_AUTH && isTruthy(process.env.SERVICE_TO_SERVICE_AUTH)) {
    return true;
  }

  return Boolean(process.env.K_SERVICE);
}

function getExpectedAudience(req) {
  const forwardedProto = req.header("x-forwarded-proto")?.split(",")[0]?.trim();
  const protocol = forwardedProto || (req.secure ? "https" : "http");
  const host = req.header("host");

  return host ? `${protocol}://${host}` : null;
}

function getBearerToken(req) {
  const authorization = req.header("authorization");
  if (!authorization) {
    return null;
  }

  const [scheme, token] = authorization.split(" ");
  if (scheme?.toLowerCase() !== "bearer" || !token) {
    return null;
  }

  return token.trim();
}

export async function requireServiceCaller(req) {
  if (!shouldRequireServiceAuth() || req.path === "/healthz") {
    return null;
  }

  const token = getBearerToken(req);
  if (!token) {
    throw createHttpError(401, "Missing service identity");
  }

  const audience = getExpectedAudience(req);
  if (!audience) {
    throw createHttpError(401, "Missing service audience");
  }

  try {
    const ticket = await authClient.verifyIdToken({
      idToken: token,
      audience
    });

    const payload = ticket.getPayload();
    if (!payload) {
      throw createHttpError(401, "Invalid service identity");
    }

    req.serviceCaller = {
      email: payload.email ?? null,
      subject: payload.sub ?? null,
      audience: payload.aud ?? null
    };

    return req.serviceCaller;
  } catch {
    throw createHttpError(401, "Invalid service identity");
  }
}
